From 39bd743024f5f40a1b21a64ef2f5f47744df7746 Mon Sep 17 00:00:00 2001
From: nkraetzschmar <9020053+nkraetzschmar@users.noreply.github.com>
Date: Thu, 19 Feb 2026 15:43:16 +0100
Subject: [PATCH] fallback to send/sendmsg syscalls if vmsplice returns -EPERM

this is useful in cases where vmsplice is blocked by seccomp filters, e.g. in containers

Signed-off-by: Stephan Mueller <smueller@chronox.de>
---
 lib/internal.h   |  1 +
 lib/kcapi-aead.c | 39 ++++++++++++++++++++++++++++++---------
 lib/kcapi-asym.c | 39 ++++++++++++++++++++++++++++++---------
 lib/kcapi-md.c   | 13 +++++++++----
 lib/kcapi-sym.c  | 39 ++++++++++++++++++++++++++++++---------
 5 files changed, 100 insertions(+), 31 deletions(-)

diff --git a/lib/internal.h b/lib/internal.h
index 43c967b..ad6a55d 100644
--- a/lib/internal.h
+++ b/lib/internal.h
@@ -259,6 +259,7 @@ struct kcapi_handle {
 	struct kcapi_aead_data aead;
 	struct kcapi_aio aio;
 	struct kcapi_flags flags;
+	unsigned int vmsplice_eperm : 1;
 };
 
 /************************************************************
diff --git a/lib/kcapi-aead.c b/lib/kcapi-aead.c
index b52dda0..76ee64a 100644
--- a/lib/kcapi-aead.c
+++ b/lib/kcapi-aead.c
@@ -449,11 +449,21 @@ IMPL_SYMVER(aead_stream_update, "1.3.1")
 ssize_t impl_aead_stream_update(struct kcapi_handle *handle,
 				struct iovec *iov, size_t iovlen)
 {
-	if (handle->processed_sg <= handle->flags.alg_max_pages)
-		return _kcapi_common_vmsplice_iov(handle, iov, iovlen,
-						  SPLICE_F_MORE);
-	else
-		return _kcapi_common_send_data(handle, iov, iovlen, MSG_MORE);
+	ssize_t ret = 0;
+
+	switch (handle->vmsplice_eperm) {
+	case 0:
+		if (handle->processed_sg <= handle->flags.alg_max_pages) {
+			ret = _kcapi_common_vmsplice_iov(handle, iov, iovlen, SPLICE_F_MORE);
+			if (ret != -EPERM) break;
+			handle->vmsplice_eperm = 1;
+		}
+		__attribute__((fallthrough));
+	default:
+		ret = _kcapi_common_send_data(handle, iov, iovlen, MSG_MORE);
+	}
+
+	return ret;
 }
 
 ORIG_SYMVER(aead_stream_update, "0.12.0")
@@ -467,10 +477,21 @@ IMPL_SYMVER(aead_stream_update_last, "1.3.1")
 ssize_t impl_aead_stream_update_last(struct kcapi_handle *handle,
 				     struct iovec *iov, size_t iovlen)
 {
-	if (handle->processed_sg <= handle->flags.alg_max_pages)
-		return _kcapi_common_vmsplice_iov(handle, iov, iovlen, 0);
-	else
-		return _kcapi_common_send_data(handle, iov, iovlen, 0);
+	ssize_t ret = 0;
+
+	switch (handle->vmsplice_eperm) {
+	case 0:
+		if (handle->processed_sg <= handle->flags.alg_max_pages) {
+			ret = _kcapi_common_vmsplice_iov(handle, iov, iovlen, 0);
+			if (ret != -EPERM) break;
+			handle->vmsplice_eperm = 1;
+		}
+		__attribute__((fallthrough));
+	default:
+		ret = _kcapi_common_send_data(handle, iov, iovlen, 0);
+	}
+
+	return ret;
 }
 
 ORIG_SYMVER(aead_stream_update_last, "0.12.0")
diff --git a/lib/kcapi-asym.c b/lib/kcapi-asym.c
index 4159478..91ec5e7 100644
--- a/lib/kcapi-asym.c
+++ b/lib/kcapi-asym.c
@@ -188,11 +188,21 @@ IMPL_SYMVER(akcipher_stream_update, "1.3.1")
 ssize_t impl_akcipher_stream_update(struct kcapi_handle *handle,
 				    struct iovec *iov, size_t iovlen)
 {
-	if (handle->processed_sg < handle->flags.alg_max_pages)
-		return _kcapi_common_vmsplice_iov(handle, iov, iovlen,
-						  SPLICE_F_MORE);
-	else
-		return _kcapi_common_send_data(handle, iov, iovlen, MSG_MORE);
+	ssize_t ret = 0;
+
+	switch (handle->vmsplice_eperm) {
+	case 0:
+		if (handle->processed_sg < handle->flags.alg_max_pages) {
+			ret = _kcapi_common_vmsplice_iov(handle, iov, iovlen, SPLICE_F_MORE);
+			if (ret != -EPERM) break;
+			handle->vmsplice_eperm = 1;
+		}
+		__attribute__((fallthrough));
+	default:
+		ret = _kcapi_common_send_data(handle, iov, iovlen, MSG_MORE);
+	}
+
+	return ret;
 }
 
 ORIG_SYMVER(akcipher_stream_update, "0.12.0")
@@ -206,10 +216,21 @@ IMPL_SYMVER(akcipher_stream_update_last, "1.3.1")
 ssize_t impl_akcipher_stream_update_last(struct kcapi_handle *handle,
 					 struct iovec *iov, size_t iovlen)
 {
-	if (handle->processed_sg < handle->flags.alg_max_pages)
-		return _kcapi_common_vmsplice_iov(handle, iov, iovlen, 0);
-	else
-		return _kcapi_common_send_data(handle, iov, iovlen, 0);
+	ssize_t ret = 0;
+
+	switch (handle->vmsplice_eperm) {
+	case 0:
+		if (handle->processed_sg < handle->flags.alg_max_pages) {
+			ret = _kcapi_common_vmsplice_iov(handle, iov, iovlen, 0);
+			if (ret != -EPERM) break;
+			handle->vmsplice_eperm = 1;
+		}
+		__attribute__((fallthrough));
+	default:
+		ret = _kcapi_common_send_data(handle, iov, iovlen, 0);
+	}
+
+	return ret;
 }
 
 ORIG_SYMVER(akcipher_stream_update_last, "0.12.0")
diff --git a/lib/kcapi-md.c b/lib/kcapi-md.c
index bddd76b..2edfd8a 100644
--- a/lib/kcapi-md.c
+++ b/lib/kcapi-md.c
@@ -46,15 +46,20 @@ static inline ssize_t _kcapi_md_update(struct kcapi_handle *handle,
 	if (len > INT_MAX)
 		return -EMSGSIZE;
 
+	switch (handle->vmsplice_eperm) {
+	case 0:
+		if (len >= (1<<15)) {
+			ret = _kcapi_common_vmsplice_chunk(handle, buffer, len, SPLICE_F_MORE);
+			if (ret != -EPERM) break;
+			handle->vmsplice_eperm = 1;
+		}
+		__attribute__((fallthrough));
+	default:
 	/* zero buffer length cannot be handled via splice */
-	if (len < (1<<15)) {
 		ret = _kcapi_common_accept(handle);
 		if (ret)
 			return ret;
 		ret = send(*_kcapi_get_opfd(handle), buffer, len, MSG_MORE);
-	} else {
-		ret = _kcapi_common_vmsplice_chunk(handle, buffer, len,
-						   SPLICE_F_MORE);
 	}
 
 	if (ret < 0)
diff --git a/lib/kcapi-sym.c b/lib/kcapi-sym.c
index 911ec1e..30295c3 100644
--- a/lib/kcapi-sym.c
+++ b/lib/kcapi-sym.c
@@ -229,11 +229,21 @@ IMPL_SYMVER(cipher_stream_update, "1.3.1")
 ssize_t impl_cipher_stream_update(struct kcapi_handle *handle,
 				  struct iovec *iov, size_t iovlen)
 {
-	if (handle->processed_sg <= handle->flags.alg_max_pages)
-		return _kcapi_common_vmsplice_iov(handle, iov, iovlen,
-						  SPLICE_F_MORE);
-	else
-		return _kcapi_common_send_data(handle, iov, iovlen, MSG_MORE);
+	ssize_t ret = 0;
+
+	switch (handle->vmsplice_eperm) {
+	case 0:
+		if (handle->processed_sg <= handle->flags.alg_max_pages) {
+			ret = _kcapi_common_vmsplice_iov(handle, iov, iovlen, SPLICE_F_MORE);
+			if (ret != -EPERM) break;
+			handle->vmsplice_eperm = 1;
+		}
+		__attribute__((fallthrough));
+	default:
+		ret = _kcapi_common_send_data(handle, iov, iovlen, MSG_MORE);
+	}
+
+	return ret;
 }
 
 ORIG_SYMVER(cipher_stream_update, "0.12.0")
@@ -247,10 +257,21 @@ IMPL_SYMVER(cipher_stream_update_last, "1.3.1")
 ssize_t impl_cipher_stream_update_last(struct kcapi_handle *handle,
 				       struct iovec *iov, size_t iovlen)
 {
-	if (handle->processed_sg <= handle->flags.alg_max_pages)
-		return _kcapi_common_vmsplice_iov(handle, iov, iovlen, 0);
-	else
-		return _kcapi_common_send_data(handle, iov, iovlen, 0);
+	ssize_t ret = 0;
+
+	switch (handle->vmsplice_eperm) {
+	case 0:
+		if (handle->processed_sg <= handle->flags.alg_max_pages) {
+			ret = _kcapi_common_vmsplice_iov(handle, iov, iovlen, 0);
+			if (ret != -EPERM) break;
+			handle->vmsplice_eperm = 1;
+		}
+		__attribute__((fallthrough));
+	default:
+		ret = _kcapi_common_send_data(handle, iov, iovlen, 0);
+	}
+
+	return ret;
 }
 
 ORIG_SYMVER(cipher_stream_update_last, "1.2.0")
-- 
2.50.1 (Apple Git-155)

